src/providers/cloudstack/configdrive.rs

```diff
@@ -69,36 +82,67 @@ impl ConfigDrive {
-    mount::mount(Some(source), target, Some(fstype), mount::MS_RDONLY, None::<&str>)
-        .chain_err(|| format!("failed to read-only mount source '{:?}' to target '{:?}' with filetype '{}'", source, target, fstype))
+    fn unmount(target: &Path) -> Result<()> {
+        mount::umount(target)
+            .chain_err(|| format!("failed to unmount target '{:?}'", target))
+    }
 }
 
-fn unmount(target: &Path) -> Result<()> {
-    mount::umount(target)
-        .chain_err(|| format!("failed to unmount target '{:?}'", target))
+impl MetadataProvider for ConfigDrive {
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        let mut out = HashMap::new();
+        let add_value = |map: &mut HashMap<_, _>, key: &str, name| -> Result<()> {
+            let value = self.fetch_value(name)?;
+
+            if let Some(value) = value {
+                map.insert(key.to_string(), value);
+            }
+
+            Ok(())
+        };
+
+        add_value(&mut out, "CLOUDSTACK_AVAILABILITY_ZONE", "availability_zone")?;
+        add_value(&mut out, "CLOUDSTACK_CLOUD_IDENTIFIER", "cloud_identifier")?;
+        add_value(&mut out, "CLOUDSTACK_INSTANCE_ID", "instance_id")?;
+        add_value(&mut out, "CLOUDSTACK_LOCAL_HOSTNAME", "local_hostname")?;
+        add_value(&mut out, "CLOUDSTACK_SERVICE_OFFERING", "service_offering")?;
+        add_value(&mut out, "CLOUDSTACK_VM_ID", "vm_id")?;
+
+        Ok(out)
+    }
+
+    fn hostname(&self) -> Result<Option<String>> {
+        Ok(None)
+    }
+
+    fn ssh_keys(&self) -> Result<Vec<AuthorizedKeyEntry>> {
+        let keys = self.fetch_publickeys()?
+            .into_iter()
+            .map(|key| AuthorizedKeyEntry::Valid{key})
+            .collect::<Vec<_>>();
+
+        Ok(keys)
+    }
+
+    fn networks(&self) -> Result<Vec<network::Interface>> {
+        Ok(vec![])
+    }
+
+    fn network_devices(&self) -> Result<Vec<network::Device>> {
+        Ok(vec![])
+    }
 }
 
-pub fn fetch_metadata() -> Result<Metadata> {
-    let drive = ConfigDrive::new()?;
-
-    Ok(Metadata::builder()
-       .add_publickeys(drive.fetch_publickeys()?)
-       .add_attribute_if_exists("CLOUDSTACK_AVAILABILITY_ZONE".into(), drive.fetch_value("availability_zone")?)
-       .add_attribute_if_exists("CLOUDSTACK_INSTANCE_ID".into(), drive.fetch_value("instance_id")?)
-       .add_attribute_if_exists("CLOUDSTACK_SERVICE_OFFERING".into(), drive.fetch_value("service_offering")?)
-       .add_attribute_if_exists("CLOUDSTACK_CLOUD_IDENTIFIER".into(), drive.fetch_value("cloud_identifier")?)
-       .add_attribute_if_exists("CLOUDSTACK_LOCAL_HOSTNAME".into(), drive.fetch_value("local_hostname")?)
-       .add_attribute_if_exists("CLOUDSTACK_VM_ID".into(), drive.fetch_value("vm_id")?)
-       .build())
+impl ::std::ops::Drop for ConfigDrive {
+    fn drop(&mut self) {
+        if self.temp_dir.is_some() {
+            ConfigDrive::unmount(&self.path).unwrap();
```

Comment:
I'm not sure we should panic here. (This is carried-over code, feel free to defer this)